src/core/server/elasticsearch/elasticsearch_service.ts

```diff
@@ -0,0 +1,122 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ConnectableObservable, Observable, Subscription } from 'rxjs';
+import { filter, map, publishReplay, switchMap } from 'rxjs/operators';
+import { CoreContext, CoreService } from '../../types';
+import { Logger } from '../logging';
+import { ClusterClient } from './cluster_client';
+import { ElasticsearchClientConfig } from './elasticsearch_client_config';
+import { ElasticsearchConfig } from './elasticsearch_config';
+
+interface CoreClusterClients {
+  config: ElasticsearchConfig;
+  adminClient: ClusterClient;
+  dataClient: ClusterClient;
+}
+
+/** @internal */
+export interface ElasticsearchServiceStartContract {
+  // Required for the BWC only.
+  readonly bwc: {
+    readonly config$: Observable<ElasticsearchConfig>;
+  };
+
+  readonly createClient: (type: string, config: ElasticsearchClientConfig) => ClusterClient;
+  readonly adminClient$: Observable<ClusterClient>;
+  readonly dataClient$: Observable<ClusterClient>;
+}
+
+/** @internal */
+export class ElasticsearchService implements CoreService<ElasticsearchServiceStartContract> {
+  private readonly log: Logger;
+  private subscription?: Subscription;
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.log = coreContext.logger.get('elasticsearch-service');
+  }
+
+  public async start(): Promise<ElasticsearchServiceStartContract> {
+    this.log.debug('Starting elasticsearch service');
+
+    const clients$ = this.coreContext.configService
+      .atPath('elasticsearch', ElasticsearchConfig)
+      .pipe(
+        filter(() => {
+          if (this.subscription !== undefined) {
+            this.log.error('Clients cannot be changed after they are created');
+            return false;
+          }
+
+          return true;
+        }),
+        switchMap(
+          config =>
+            new Observable<CoreClusterClients>(subscriber => {
+              this.log.debug(`Creating elasticsearch clients`);
+
+              const coreClients = {
+                config,
+                adminClient: this.createClusterClient('admin', config),
+                dataClient: this.createClusterClient('data', config),
+              };
+
+              subscriber.next(coreClients);
+
+              return () => {
+                this.log.debug(`Closing elasticsearch clients`);
+
+                coreClients.adminClient.close();
+                coreClients.dataClient.close();
+              };
+            })
+        ),
+        publishReplay(1)
+      ) as ConnectableObservable<CoreClusterClients>;
+
+    this.subscription = clients$.connect();
```

Comment:
So, it's the easiest way I can think of to fulfill our requirements:
* Don't resubscribe to source config and admin/data client creation observable ("share" part);
* Unsubscribe from source observable (and close data and admin clients) when we stop the service ("dispose" part)

We can't use `shareReplay` here because it [won't unsubscribe from source observable](https://github.com/ReactiveX/rxjs/issues/3336) (and hence won't close clients) on unsubscribe. It does that only if source observable errors/completes that isn't the case here.

And we can't use `publishReplay(1).refCount()` since it will unsubscribe from source observable in non-controlled way i.e. when ref count goes from 1 to 0.

We can try to use `ReplaySubject` directly here instead, but we'll have more code. I need to add clarifying comment here either way though. 

What do you think?

Reply:
Makes sense

Reply:
Good to know! We may change implementation here once we upgrade to newer RxJS.